src/coreclr/jit/codegenxarch.cpp

```diff
@@ -6074,17 +6074,31 @@ void CodeGen::genCall(GenTreeCall* call)
     }
 #endif // defined(DEBUG) && defined(TARGET_X86)
 
-    // When it's a PInvoke call and the call type is USER function, we issue VZEROUPPER here
-    // if the function contains 256bit AVX instructions, this is to avoid AVX-256 to Legacy SSE
-    // transition penalty, assuming the user function contains legacy SSE instruction.
-    // To limit code size increase impact: we only issue VZEROUPPER before PInvoke call, not issue
-    // VZEROUPPER after PInvoke call because transition penalty from legacy SSE to AVX only happens
-    // when there's preceding 256-bit AVX to legacy SSE transition penalty.
-    // This applies to 512bit AVX512 instructions as well.
-    if (call->IsPInvoke() && (call->gtCallType == CT_USER_FUNC) && (GetEmitter()->Contains256bitOrMoreAVX()))
-    {
-        assert(compiler->canUseVexEncoding());
-        instGen(INS_vzeroupper);
+    if (compiler->canUseVexEncoding())
+    {
+        // The Intel optimization manual guidance in `3.11.5.3 Fixing Instruction Slowdowns` states:
+        //   Insert a VZEROUPPER to tell the hardware that the state of the higher registers is clean
+        //   between the VEX and the legacy SSE instructions. Often the best way to do this is to insert a
+        //   VZEROUPPER before returning from any function that uses VEX (that does not produce a VEX
+        //   register) and before any call to an unknown function.
+
+        bool needsZeroupper = false;
+
+        // TODO-XArch-CQ: We should emit vzeroupper before R2R calls if they aren't known to require AVX+
```

Comment:
I do not think we need to worry about the transient R2R use with tiered JIT. The only interesting R2R case is disabled tiered JIT. We would need a new JIT/EE interface flag to tell the JIT whether the tiered JIT is globally disabled for the whole process.

Reply:
👍
 
My main concern here is how that can negatively impact startup perf. The perf penalty here can be up to a 10x throughput regression per instruction (it's still around 4x on the latest hardware).

If we can't do this here, it might be interesting to see if it can be done as part of one of the stubs or even just have the R2R compiler emit it as a `if (cachedAvxSupportedCheck) { vzeroupper }` in the prologue of a R2R method that uses floating-point instructions.

Reply:
I think doing this in the stubs would make the most sense.